### DISCUSSION TITLE
How to add nodes to graphs in loaded dataset

### DISCUSSION THREAD

**Discussion by simon-forb** [pyg-team/pytorch_geometric#8924] [Q&A] [answered]
What would be the most canonical way to add nodes to a dataset that is imported through `torch_geometric`?

For instance, I am loading the reddit binary dataset like this:

```python
reddit = torch_geometric.datasets.TUDataset("/tmp/reddit-b", name="REDDIT-BINARY")
```

This yields a dataset which contains 2000 graphs. Now I would like to add a number of nodes to each graph. I only figured out how to do this using `networkx` but I believe there is probably a better way to do this without this double conversion.

**Comment by rusty1s** [comment, marked as answer]
You can modify graphs in `Dataset` through the `transform`/`pre_transform` interface. For how to add new nodes, take a look at, e.g., the [`VirtualNode`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/transforms/virtual_node.py) transform. In general, it is as simple as
```python
data.edge_index = torch.cat([data.edge_index, torch.tensor(new_edges)], dim=1)
data.x = torch.cat([data.x, new_features], dim=0)
